src/CoreWCF.Http/src/CoreWCF/Channels/HttpChannelHelpers.cs

```diff
@@ -121,74 +123,81 @@ private void ThrowMaxReceivedMessageSizeExceeded()
             }
         }
 
-        private async Task<Message> DecodeBufferedMessageAsync(ArraySegment<byte> buffer, Stream inputStream)
+        private async ValueTask<Message> DecodeBufferedMessageAsync(ReadOnlySequence<byte> buffer)
         {
             try
             {
-                // if we're chunked, make sure we've consumed the whole body
-                if (ContentLength == -1 && buffer.Count == _settings.MaxReceivedMessageSize)
+                // if we're chunked, ensure message size is lower than or equal to MaxReceivedMessageSize
+                if (ContentLength == -1 && buffer.Length > _settings.MaxReceivedMessageSize)
                 {
-                    byte[] extraBuffer = new byte[1];
-                    int extraReceived = await inputStream.ReadAsync(extraBuffer, 0, 1);
-                    if (extraReceived > 0)
-                    {
-                        ThrowMaxReceivedMessageSizeExceeded();
-                    }
+                    ThrowMaxReceivedMessageSizeExceeded();
                 }
 
-                try
-                {
-                    return _messageEncoder.ReadMessage(buffer, _bufferManager, ContentType);
-                }
-                catch (XmlException xmlException)
-                {
-                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
-                        new ProtocolException(SR.MessageXmlProtocolError, xmlException));
-                }
+                return await _messageEncoder.ReadMessageAsync(buffer, _bufferManager, ContentType);
             }
-            finally
+            catch (XmlException xmlException)
             {
-                inputStream.Close();
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                    new ProtocolException(SR.MessageXmlProtocolError, xmlException));
             }
         }
 
         private async Task<Message> ReadBufferedMessageAsync(Stream inputStream)
         {
-            ArraySegment<byte> messageBuffer = GetMessageBuffer();
-            byte[] buffer = messageBuffer.Array;
-            int offset = 0;
-            int count = messageBuffer.Count;
-
-            while (count > 0)
+            var reader = PipeReader.Create(inputStream);
+            try
             {
-                int bytesRead = await inputStream.ReadAsync(buffer, offset, count);
-                if (bytesRead == 0) // EOF
+                ReadOnlySequence<byte> buffer;
+                while (true)
                 {
-                    if (ContentLength != -1)
+                    ReadResult result = await reader.ReadAsync();
+                    buffer = result.Buffer;
+
+                    reader.AdvanceTo(buffer.Start, buffer.End);
+
```

Comment:
I presume the idea here is to force it to keep on reading without consuming anything until the entire request has been read. The [documentation](https://learn.microsoft.com/en-us/dotnet/api/system.io.pipelines.pipereader.advanceto?view=dotnet-plat-ext-8.0#system-io-pipelines-pipereader-advanceto(system-sequenceposition-system-sequenceposition)) for PipeReader.AdvanceTo explicitly states "The Buffer previously returned from ReadAsync(CancellationToken) must not be accessed after this call.". While currently you're getting away with it due to using PipeReader.Create, when we move on to .NET 6 or later, we won't be getting the body as a Stream, we'll get it as a PipeReader and that might not work. You should move the AdvanceTo to be after the check for IsCompleted and then call AdvanceTo in the finally block before calling CompleteAsync. This way the returned buffer is guaranteed to be valid when calling DecodeBufferedMessageAsync.

Reply:
Thanks. done [here](https://github.com/CoreWCF/CoreWCF/pull/801/commits/47b8404eb568804204b09fa1a3d5a5de9beec845)
I had to add an additional commit to avoid the buffers from PipeReader being freed before DispatchAsync is called . please see https://github.com/CoreWCF/CoreWCF/pull/801/commits/58f962a8249494517772a305c3fbb18a3b851228